### PR TITLE
ARROW-9096: [Python] Pandas roundtrip with dtype="object" underlying numeric column index

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes ARROW-9096
- [x] tests added & passed

This PR fixes the roundtrip conversion for Pandas DataFrames whose column index is numeric but has `dtype=object`, such as
```
df = pd.DataFrame([1], columns=pd.Index([1], dtype=object))  # underlying int
df = pd.DataFrame([1], columns=pd.Index([1.1], dtype=object)) # underlying float
df = pd.DataFrame([1], columns=pd.Index([datetime(2018, 1, 1)], dtype='object')) # underlying datetime
```
https://issues.apache.org/jira/browse/ARROW-3651 largely solved the datetime variant of this problem (such that the conversion ran correctly excepting that the dtype after roundtrip did not match). With the current fix a roundtrip of the problematic DataFrames from ARROW-3651 returns the exact original frame.

### CODE REVIEW

File: python/pyarrow/pandas_compat.py [hunk 1]
```diff
@@ -1009,12 +1009,15 @@ def _is_generated_index_name(name):
     return re.match(pattern, name) is not None
 
 
+# ARROW-9096: added integer and floating
```
Comment (emkornfield) [inline python/pyarrow/pandas_compat.py:1012]: nit: generally only have comment like this for TODOs.  git blame/git log can track when things were added.

File: python/pyarrow/pandas_compat.py [hunk 2]
```diff
@@ -1080,8 +1083,10 @@ def _reconstruct_columns_from_metadata(columns, column_indexes):
     ]
 
     # Convert each level to the dtype provided in the metadata
+    # ARROW-9096: need numpy_type to match cast against original DataFrame
```
Comment (emkornfield) [inline python/pyarrow/pandas_compat.py:1086]: nit: drop JIRA reference.  maybe also the whole comment since this is used directly below?

**Comment by emkornfield** [comment]
CC @jorisvandenbossche

File: python/pyarrow/pandas_compat.py (hunk 2 above)
Comment (arw2019) [reply python/pyarrow/pandas_compat.py]: makes sense - it's easy to see what's going on w/o this line so dropped it as suggested

**Review by wesm — APPROVED** [review]
+1

**Comment by arw2019** [comment]
thanks @wesm @emkornfield for reviewing!!

**Pull request closed without merging**